### PR TITLE
[office-js] Making AsyncResult.value generic

### PR DESCRIPTION
**Pull request by AlexJerabek**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://dev.office.com/reference/add-ins/shared/asyncresult
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Review by ElizabethSamuel-MSFT — APPROVED** [review]

**Comment by AlexJerabek** [comment]
DT Maintainers: Still waiting on a review from @Zlatkovsky. Please do not merge until then.
Thanks!

File: types/office-js/index.d.ts
Comment (Rick-Kirkham) [inline types/office-js/index.d.ts]: Shouldn't this be `AsyncResult<string>` instead of` AsyncResult<any>`? AsyncResult.value is defined at all, its going to be an access token; hence, a string.

File: types/office-js/index.d.ts
Comment (ElizabethSamuel-MSFT) [inline types/office-js/index.d.ts]: nit: these comments should end with a period.

File: types/office-js/index.d.ts
Comment (Rick-Kirkham) [inline types/office-js/index.d.ts]: Why undefined? Do we know AsyncResult.value is always undefined when there's a callback to removeHandlerAsync?

**Review by Rick-Kirkham — APPROVED** [review]
I'll approve but please take a look at my comments (more questions than comments I guess).

File: types/office-js/index.d.ts
Comment (Zlatkovsky) [inline types/office-js/index.d.ts]: Since it's not anticipated to ever return a value, it should be `AsyncResult<void>` instead

**Review by Zlatkovsky — APPROVED** [review]
Reviewed and left some comments on the Shared APIs.
Will leave the Outlook ones to Elizabeth

File: types/office-js/index.d.ts
Comment (AlexJerabek) [reply types/office-js/index.d.ts]: Won't this affect existing code? Or can people ignore the generics?

File: types/office-js/index.d.ts
Comment (AlexJerabek) [reply types/office-js/index.d.ts]: API Extractor/Documenter need the Office prefix to autolink in the reference docs. I'm not if that feature is smart enough to work here, but for the future...

File: types/office-js/index.d.ts
Comment (AlexJerabek) [reply types/office-js/index.d.ts]: Changed to getSelectedDataAsync<T> … result: AsyncResult<T>, as per your earlier comment.

**Comment by AlexJerabek** [comment]
Responded to your feedback @Zlatkovsky. Thank you.

Follow up question, should we change more of the AsyncResult<any> to function<T>(AsyncResult<T>)?

**Comment by AlexJerabek** [comment]
DT Maintainers: This should be good to merge now. Thanks!

**Pull request merged**
